### PR TITLE
Update asynchttpclient-2.0 to Instrumenter API

### PR DESCRIPTION
**Pull request by trask**

### CODE REVIEW

File: instrumentation/async-http-client/async-http-client-1.9/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v1_9/ResponseInstrumentation.java [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-package io.opentelemetry.javaagent.instrumentation.asynchttpclient;
+package io.opentelemetry.javaagent.instrumentation.v1_9;
```
Comment (mateuszrzeszutek) [inline instrumentation/async-http-client/async-http-client-1.9/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v1_9/ResponseInstrumentation.java:6]: It probably should be `...asynchttpclient.v1_9`

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncCompletionHandlerInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.v2_0;
```
Comment (mateuszrzeszutek) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncCompletionHandlerInstrumentation.java:6]:
```suggestion
package io.opentelemetry.javaagent.instrumentation.asynchttpclient.v2_0;
```

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java [hunk 3]
```diff
@@ -38,24 +36,38 @@ public void transform(TypeTransformer transformer) {
...
+  public static class ExecuteRequestAdvice {
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
     public static void onEnter(
         @Advice.Argument(0) Request request,
         @Advice.Argument(1) AsyncHandler<?> handler,
         @Advice.Local("otelScope") Scope scope) {
       Context parentContext = currentContext();
-      if (!tracer().shouldStartSpan(parentContext)) {
+      if (!AsyncHttpClientSingletons.instrumenter().shouldStart(parentContext, request)) {
```
Comment (mateuszrzeszutek) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java:50]: Static import?

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java [hunk 4]
```diff
@@ -38,24 +36,38 @@ public void transform(TypeTransformer transformer) {
...
+      // (similar to other http client instrumentations, and needed for library instrumentation)
+      //
+      // when doing this, note that AsyncHttpClient has different behavior if the AsyncHandler also
+      // implements ProgressAsyncHandler or StreamedAsyncHandler (or both)
+      // so four wrappers are needed to match the different combinations so that the wrapper won't
+      // affect the behavior
+      //
+      // when doing this, also note that there was a breaking change in AsyncHandler between 2.0 and
+      // 2.1, so the instrumentation module will need to be essentially duplicated (or a common
+      // module introduced)
```
Comment (mateuszrzeszutek) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java:67]: 👍

File: instrumentation/async-http-client/async-http-client-1.9/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v1_9/ResponseInstrumentation.java (hunk 1 above)
Comment (trask) [reply instrumentation/async-http-client/async-http-client-1.9/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v1_9/ResponseInstrumentation.java]: 🤦‍♂️

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java (hunk 3 above)
Comment (trask) [reply instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/v2_0/AsyncHttpClientInstrumentation.java]: ugh Intellij 👍

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientNetAttributesExtractor.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+final class AsyncHttpClientNetAttributesExtractor
+    extends NetAttributesExtractor<Request, Response> {
+
+  @Override
+  public String transport(Request request) {
+    return SemanticAttributes.NetTransportValues.IP_TCP;
+  }
+
+  @Override
+  public @Nullable String peerName(Request request, @Nullable Response response) {
```
Comment (anuraaga) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientNetAttributesExtractor.java:23]: Is this nullable?

**Pull request merged**